### PR TITLE
Add dockerfile

### PR DESCRIPTION
**Pull request by skrider**
Implements #1293 

Adds a dockerfile to run vLLM in a containerized environment. This has been tested to work with `facebook/opt-13b` on 2 `g5.2xlarge` AWS instances utilizing Ray for communication. Output of `benchmarks/benchmark_serving.py`:

```
Total time: 16.32 s
Throughput: 61.28 requests/s
Average latency: 10.59 s
Average latency per token: 0.88 s
Average latency per output token: 2.12 s
```

The image is quite large however, at 6.93 gb.

Is this testing setup sufficient?

Additionally, using `cuda-base` reduces the image size significantly but prevents installation of development dependencies,  among other things preventing unit tests from being run - is this a concern?

Should we also consider pruning the image and removing unused dependencies from the pip install to further reduce image size?

### CODE REVIEW

**Comment by simon-mo** [comment]
Hi @skrider

Great work so far. This is a good start. Few comments:
* We should absolutely try to prune the image as much as possible. 
* Can we add an `ENTRYPOINT` so users don't need to type `python3 -m vllm.entrypoints.api_server` each time.
* Consider add `EXPOSE` for default port as well?
* Ideally, we should have "dev" container that contains all build dependencies so we can run CI and help contributors set up everything; and another "prod" container with smallest size possible to run models or built upon.

**Comment by skrider** [comment]
@simon-mo - noted, will do. I am working on some tooling to help prune images automatically here: https://github.com/skrider/pip-prune . So far have been able to reduce pip install footprint from 6 to 3 gb.

Regarding entrypoint - I considered adding this but thought that sometimes users might want to start the container with ray and then exec `vllm.entrypoints.api_server` later. Perhaps it would be good to provide an `entrypoint.sh` script that checks for an environment variable and optionally starts up/waits for Ray? Could get a bit clunky. Or should we just ignore Ray?

Regarding the dev/prod container - should these install vllm from pypi or the repo directly?

**Comment by simon-mo** [comment]
* Users can override entrypoint using `--entrypoint` command in `docker run` as well as similar means in K8s. 
* It would be great to allow customizing version using docker `ARG` and default to the latest version. As long as it is easy enough to build it for any version, it should be fine.

File: Dockerfile
Comment (NikolaBorisov) [inline Dockerfile:23]: Ideally here we will copy only csrc folder and build only the c++ code. If you copy vllm folder any changes to the python code causes slow rebuild of the c++ code, most of the time this is not needed.

Also the README.md file is kind of required during the build of the c++ code, but we can copy empty README.me while build the c++ code.

File: Dockerfile
Comment (NikolaBorisov) [inline Dockerfile:43]: First we should install the requirements. They change rarely. This way there will be large layer with all the pip packages. And the vllm code will be in another layer. For example if you modify the vllm code and build the docker and push to some repo, the amount that is pushed will be small.

I think we should copy the vllm folder from source and the *.so files

File: Dockerfile
Comment (NikolaBorisov) [inline Dockerfile:48]: I'm not sure but I think many people use the openai api server. Ideally we can add another entry point for it. That server seem to have some requirements.

**Review by NikolaBorisov — COMMENTED** [review]
Left few comments. Trying to reconcile this with #1415

**Comment by NikolaBorisov** [comment]
>     * It would be great to allow customizing version using docker `ARG` and default to the latest version. As long as it is easy enough to build it for any version, it should be fine.

I don't think we need this. Users who want particular version would be able to just `docker pull vllm:0.2.0` If they want to build from scratch they can checkout the right branch/tag and docker build there.

File: Dockerfile
Comment (skrider) [reply Dockerfile:23]: Makes sense. Do you think it would be possible to build the extensions separate from the wheel, so that the wheel building step only bundles everything without having to rebuild?

Comment (NikolaBorisov) [reply Dockerfile:23]: I'm not sure how to build the wheel with already build extensions, but it might be possible. I don't think we need to build the wheel. If you want we can do another container path for building the wheel if you want this to be used to publish the wheel files to pip?

Comment (NikolaBorisov) [reply Dockerfile:23]: Ok I just saw you did another stage to build the wheel. This is ok I think.

File: Dockerfile
Comment (NikolaBorisov) [inline Dockerfile:33]: I would replace this line with CMD touch README.md

Comment (skrider) [reply Dockerfile:33]: Think it is odd that the build depends on documentation - makes sense for the release but given the amount of time it takes to build I don't think it makes sense for development builds. Think making this an optional dependency of the build process would be better

File: Dockerfile
Comment (skrider) [reply Dockerfile]: I think for now no need to build the wheel in the dockerfile, can do that later if decide to consolidate docker/CI

**Comment by skrider** [comment]
@NikolaBorisov Took your commentary into account - should be ready to merge. I am still working on pruning the image but that is not critical.

**Review by NikolaBorisov — APPROVED** [review]
Looks good to me. Just left 2 more suggestions.

**Comment by NikolaBorisov** [comment]
@simon-mo I think this is ready to merge after @skrider fixes 2 small things. I'm going to close #1415 in favor of this one.

The only think other thing we might want to fix is the names of the last to stages in the file. Instead of 'api_server' and 'openai_api_server' I think 'vllm' and 'vllm-openai' are better names.

**Comment by simon-mo** [comment]
@NikolaBorisov thank you for your feedback! @skrider it would be great to address them. I also created `vllm` docker user and i'm planning to publish official images along with releases. Before we merge, @skrider can you add comments to Dockerfile and group different stages together so they are more maintainable for the future?

**Comment by skrider** [comment]
@simon-mo Yes will do- should be done EOD

**Comment by Extremys** [comment]
> Implements #1293
> 
> Adds a dockerfile to run vLLM in a containerized environment. This has been tested to work with `facebook/opt-13b` on 2 `g5.2xlarge` AWS instances utilizing Ray for communication. Output of `benchmarks/benchmark_serving.py`:
> 
> ```
> Total time: 16.32 s
> Throughput: 61.28 requests/s
> Average latency: 10.59 s
> Average latency per token: 0.88 s
> Average latency per output token: 2.12 s
> ```
> 
> The image is quite large however, at 6.93 gb.
> 
> Is this testing setup sufficient?
> 
> Additionally, using `cuda-base` reduces the image size significantly but prevents installation of development dependencies, among other things preventing unit tests from being run - is this a concern?
> 
> Should we also consider pruning the image and removing unused dependencies from the pip install to further reduce image size?

Did you check that is working for multiGPUs inference also? (a model loaded on at least two GPUs)
Are you building the container from the target host with nvidia runtime enabled?

File: Dockerfile
Comment (simon-mo) [inline Dockerfile:1]:
```suggestion
ARG MAX_JOBS=4

FROM nvidia/cuda:11.8.0-devel-ubuntu22.04 AS dev
```

Forward declare the argument and set sensible default

File: Dockerfile
Comment (simon-mo) [inline Dockerfile:29]:
```suggestion
ENV MAX_JOBS=$MAX_JOBS 
```

File: docs/source/serving/deploying_with_docker.rst
Comment (simon-mo) [inline docs/source/serving/deploying_with_docker.rst:10]:
```suggestion
    $ DOCKER_BUILDKIT=1 docker build . --target vllm --tag vllm --build-arg MAX_JOBS=8
```

**Review by simon-mo — APPROVED** [review]

**Review by simon-mo — APPROVED** [review]

**Pull request merged**
